crates/agglayer-node/src/rpc/error.rs

```diff
@@ -0,0 +1,157 @@
+//! Support for structured errors in RPC.
+
+use ethers::{middleware::Middleware, types::H256};
+use jsonrpsee::types::error::ErrorObjectOwned;
+use serde::Serialize;
+
+use crate::kernel::{CheckTxStatusError, SignatureVerificationError, ZkevmNodeVerificationError};
+
+/// RPC error codes.
+pub mod code {
+    /// Rollup is not registered.
+    pub const ROLLUP_NOT_REGISTERED: i32 = -10001;
+
+    /// Rollup signature verification failure.
+    pub const SIGNATURE_MISMATCH: i32 = -10002;
+
+    /// Proof or state validation failed.
+    pub const VALIDATION_FAILURE: i32 = -10003;
+
+    /// L1 settlement failure.
+    pub const SETTLEMENT_ERROR: i32 = -10004;
+
+    /// Transaction status retrieval error.
+    pub const STATUS_ERROR: i32 = -10005;
+}
+
+#[derive(PartialEq, Eq, Serialize, Debug, thiserror::Error)]
+#[serde(rename_all = "kebab-case")]
+pub enum ValidationError {
+    #[error("Dry run failed")]
+    DryRun { detail: String },
+
+    #[error("State root verification failed")]
+    RootVerification { detail: String },
+}
+
+#[derive(PartialEq, Eq, Serialize, Debug, thiserror::Error)]
+#[serde(rename_all = "kebab-case")]
+pub enum SettlementError {
```

Comment:
This is mostly due to the `Serialize` instance which is used to populate the `"data"` field of the json error. Putting `Serialize` on `kernel::SettlementError` would probably require a custom impl or heavy annotation since most of the fields are not `Serialize`. Another way would be to come up with a different way of creating the structured data but I find the `Serialize`-based approach rather neat.

Also the error messages provided by thiserror here are a bit more brief. They end up in the `"message"` field, leaving most of the detail to the `"data"` field.

In summary this approach gives us control over how the json error is rendered without compromising on information stored in `kernel::SettlementError`.